Extend qwerty support to 3 octaves +other changes

Shayan-To opened the pull request:
I got developing to extend the qwerty support, but ended up doing other things as well.

Here's a list of what this includes:

- Update project to use yarn instead of npm
- Format all files (I think prettier was updated)
- Fix an exception I encountered: the black keys were being drawn to canvas before the image was loaded
- Fix the flicker problem caused by `usePersistedState`
- Fix pressing arrow keys while a key is pressed (was a to-do in the code)
- Extend the qwerty support to 3 octaves
- Add support for Dvorak keyboards

Please review the commits individually. If any changes needed to be done by me, let me know.

Fixes #151

samouri commented:
Hey @Shayan-To, 

Thank you for the Pull Request!

- Do you mind undoing the npm to yarn conversion? Even if I were happy to do that (which I don't think I am), I'd prefer that in  a separate PR.
- What was the flicker problem with usePersistedState?
- Lets run `npm run format` in a separate PR to isolate the meaningful changes

samouri commented on package.json:
```diff
@@ -46,7 +46,7 @@
     "build": "next build",
     "lint": "next lint",
-    "format": "prettier --write **/*.{ts,tsx}",
+    "format": "prettier --write '**/*.{ts,tsx}'",
```
What does this do differently?

Shayan-To replied:
I have no idea, but I think yarn was expanding that glob and causing some problems.

The change is reverted now.

samouri commented on src/app/freeplay/components/TopBar.tsx:
```diff
@@ -26,6 +28,7 @@ export default function TopBar({
   onClickRecord,
 }: TopBarProps) {
   const recordTooltip = isRecordingAudio ? 'Stop recording' : 'Start recording audio'
+  const [isDvorak, setIsDvorak] = usePersistedState('isDvorak', false, (v) => setKeyboardConfig(v))
```
out of curiosity, do you use dvorak?

samouri replied:
Super cool! I've always wanted to try it 😄

Shayan-To commented:
Okay, I managed to do everything you asked and also fixed Copilot's comments.

> Do you mind undoing the npm to yarn conversion? Even if I were happy to do that (which I don't think I am), I'd prefer that in a separate PR.

I removed the commits.

About yarn, I've been using it for some years now. I like a few things about it over npm.

- Its output is more readable.
- It is more consistent and faster at creating `node_modules`. (It creates it from zip files from its cache.)
- You always know exactly what version of it you're running for each project (because of corepack).
- Its `patch` feature is very convenient when you get into difficult troubles.
- It seems it has very good support for mono-repos and such, though I've never been in a situation to use it.

I encourage you to try it!

> What was the flicker problem with usePersistedState?

I saw the problem in training pages. It was causing the components that used it to constantly keep rendering (because `init` was an object, and inevitably changed on every render). And the `useEffect` was meant to actually only run once. (There also was a comment on the `useEffect`, though it was not accurate.)

> Lets run `npm run format` in a separate PR to isolate the meaningful changes

I removed the commit, but then it was very difficult to manage running `npx prettier --write file1 file2`, copy and pasting paths of changed files and stuff, but I managed!

----

Please take another look into this @samouri !

ndren commented:
Hey, very cool PR, thanks! I'd love to see this feature.

Side note, maybe not the right solution here, but FYI desktop browsers let you specify keys based on position in the keyboard rather than key value, such that WASD can be used as arrow keys in QWERTY, AZERTY, DVORAK etc: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code. (It's a bit more verbose though.)

samouri commented:
> Also, a question. Why isn't package-lock.json committed? Npm is very explicit about wanting it to be committed!

Ah, I forgot. I had converted to `bun` awhile back. You'll notice a bun lockfile (`bun.lockb`).

> I saw the problem in training pages.

I'm surprised you found the training pages! I never linked to them from anywhere, as they aren't ready in the slightest. They were a fun experiment I should probably delete 😆. Either way, thank you for the fix!

> Side note, maybe not the right solution here, but FYI desktop browsers let you specify keys based on position in the keyboard rather than key value, such that WASD can be used as arrow keys in QWERTY, AZERTY, DVORAK etc

Ah! Super interesting. Looks like it is supported on a majority of platforms, with one big exception. Chrome on Android seems to be missing support. It seems like `code` was added with this specific use-case in mind, so it definitely seems appropriate to support. We can fall-back the `.key` if the `.code` isn't present.

samouri commented on src/features/midi/index.tsx:
```diff
@@ -142,30 +177,45 @@ class MidiState {
-    const key = e.key.toLowerCase()
     if (key === 'arrowup') {
-      this.octave = Math.min(7, this.octave + 1)
+      this.octaveDiff = Math.min(2, this.octaveDiff + 2)
```
Can you explain why you prefer middle c + diff, versus the single piece of state (octave). Also doesn't this `2` max mean octave 7 can't be reached? Same question for the min being `-2`, means octave 1 can't be reached.

samouri commented on src/features/drawing/piano.ts:
```diff
@@ -1,9 +1,8 @@
-import { roundRect, roundCorner } from '@/features/drawing'
 import midiState from '../midi'
 import { isPointerDown } from '../pointer'
-import { isNumber } from '@/utils'
```
I think removing `isNumber` is very breaking, but it works if you add this line back

Shayan-To replied:
This change was made by prettier! 🤷‍♂️

samouri commented:
@Shayan-To I was reviewing the keyboard mapping, and this is pretty cool! I think you've made it so we take full advantage of the keyboard via using 4 rows. I think it would even work down to a 60% keyboard. 

My one idea to improve, is I think it might be clearer if we made each row-pair (AZ, 1Q) span exactly an octave rather than the octave+2 notes. That way `Z` is a `C`, and so is `Q`. That makes the two paair perfect mirrors, and I assume that'll be easier to pick up for folks.

Something like this:

<img width="600" alt="image" src="https://github.com/user-attachments/assets/8385c90e-80fe-4501-98b6-a29e387b29ae" />

samouri commented:
@Shayan-To, fyi i have some free time today, so I plan to make the rest of the changes I outlined. feel free to respond if you have further thoughts, and thank you for the help so far!

Shayan-To commented:
Thanks for taking the time.

I have one suggestion. I think it's a good idea to add a "Keyboard Options" dialog and put Qwerty/Dvorak selection there, as well as the 2-octave / 3-octave keyboard with some preview of the keyboard so people can more easily figure out the keys.

I used the 3-octave configuration that I suggested, and I really liked the layout. F of the last octave is right above C of the middle one, and it makes finding the keys very easy.

BTW, thanks for the great project! I was looking for something like it for a very long while, and it's awesome!

samouri commented:
> I think it's a good idea to add a "Keyboard Options" dialog and put Qwerty/Dvorak selection there

@Shayan-To: why do we want dvorak/qwerty options? I tested with dvorak on my laptop, and I believe it works well now. Using [KeyboardEvent.code](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code) means we get dvorak support for free.

> as well as the 2-octave / 3-octave keyboard with some preview of the keyboard so people can more easily figure out the keys.

I think that's a great idea. We definitely should have an image preview of the mapping in the app, and making it configurable would be nice, but generally lean away from configurability. I'd rather think hard about optimal layout and only support one. I'm pretty open to changing the current mapping, since I made it in a vacuum. For example, what do DAWs and/or other apps in this space do for mappings? I'd be very surprised if any supported 3-octaves at once, but if that's actually a norm I'd be happy to switch to it. Or if you personally use this mapping and tell me its worse than the 3-octave one, that'd push me towards it too!

One thing I'm confused about: I was under the impressions that folks mostly use the keyboard mapping for very light play/testing things out, but that it isn't quite ergonomic, effective enough, or worth getting good at compared to a midi keyboard.  Do you intend to use sightread without a midi keyboard for extended periods of time? If yes, then I could see the 3-octaves being more helpful than I expected.

> BTW, thanks for the great project! I was looking for something like it for a very long while, and it's awesome!

🙇 🙏
